### ISSUE TITLE
"Async & Performance" chapter4 Delegating Recursion example

### ISSUE THREAD

**Issue by bfulop** [getify/You-Dont-Know-JS#419]
Hello Getify,

I'm testing the recursion example in your book "Async & Performance", ch4.md around line number 1490:

```js
function *foo(val) {
	if (val > 1) {
		// generator recursion
		val = yield *foo( val - 1 );
	}

	return yield request( "http://some.url/?v=" + val );
}

function *bar() {
	var r1 = yield *foo( 3 );
	console.log( r1 );
}

run( bar );
```

I've been on it for some time, but in my tests the `request` is always called with 1 as a value for `val`. I would have expected 3,2,1. Where does this logic fall? Is it the good ol' function scope at play?

And thanks for all your work on this series and lectures at Frontendmasters! I finally start to understand JavaScript!

**Comment by bfulop** [comment]
If I add `var myval = val` and use that for the `request` I get a different url on each call:

```js
function *foo(val) {
  var myval = val;
  if (myval > 1) {
    // generator recursion
    val = yield *foo( myval - 1 );
  }

  return yield request( myval );
}
```

However the results are coming back "backwards": 1,2,3

**Comment by getify** [comment]
Let's use a `request(..)` function like this:

```js
function request(url) {
	return new Promise(function(resolve){
		setTimeout(function(){
			resolve( url.match(/v=(\d+)$/)[1] );
		},1000);
	});
}
```

As you can see, whatever number is on the `v=..` part of the passed in URL is what gets returned 1 second later. So:

```js
request("...v=1")
.then(function(v){
	console.log(v); // 1
});
```

So, if you follow the recursion down the line, we start with `val` at `3`, then at `2`, then at `1`, all before any request is made. Then we make a request like `return yield request("..v=1")`. What does that do? It waits for the `1` to come back from that promise, then `return`s it.

Now, we go back out one step of the recursion, and we're at the paused `val = yield *foo(..)` call. We've returned `1`, so then we assign that to `val`. Next, we make another `request(..)` call, but `val` is still `1`, and `1` still comes back. We unravel and do that a third time, and thus we get `1` passed to `request(..)` all 3 times.

Does that make sense?

--------

Now, if `*foo(..)` instead looked like this:

```js
function *foo(val) {
	if (val > 1) {
		// generator recursion
		yield *foo( val - 1 );
	}

	return yield request( "http://some.url/?v=" + val );
}
```

The only thing I changed is to take off the `val = ` part from the `yield *` line. What's the difference here? Well, we still start with a `request("..v=1")`, but when we unroll one step of the recursion, we're back to a state where `val` is still `2`, and since we don't override it with `1`, now we get a call to `request("..v=2")`. One more step unrolled, and we get `request("..v=3")`.

In other words, we get `1,2,3` as the natural order, not the backwards order. ;-)

Hope that helps!

**Comment by getify** [comment]
BTW, in a second edition, I'll probably change `*foo(..)` to look like this:

```js
function *foo(val) {
	if (val > 1) {
		// generator recursion
		console.log( "response:", yield *foo( val - 1 ) );
	}

	return yield request( "http://some.url/?v=" + val );
}
```

Which in our scenario would print out:

```
response: 1
response: 2
response: 3
```

**Comment by bfulop** [comment]
Kyle, thanks for your explanation! Yeah I think I start to slowly understand the concept. Run to completion vs iteration... We're iterating in the `if` statement in the first three "runs".
It's a good exercise to understand the flow of the program!

**Comment by getify** [comment]
;-) great, glad that helped!
